src/interpreter/blockchain.rs

```diff
@@ -308,4 +309,64 @@ where
 
         self.inc_pc()
     }
+
+    pub(crate) fn message_output(&mut self, a: Word, b: Word, c: Word, d: Word) -> Result<(), RuntimeError> {
+        if b > self.params.max_message_data_length
+            || b > MEM_MAX_ACCESS_SIZE
+            || a > VM_MAX_RAM - b - Bytes32::LEN as Word
+        {
+            return Err(PanicReason::MemoryOverflow.into());
+        }
+
+        let idx = c;
+        let amount = d;
+
+        // Safety: checked len
+        let recipient = unsafe { Address::from_slice_unchecked(&self.memory[a as usize..a as usize + Address::LEN]) };
+        if recipient == Address::zeroed() {
+            return Err(PanicReason::ZeroedMessageOutputRecipient.into());
+        }
+
+        let offset = self
+            .tx
+            .output_offset(c as usize)
+            .map(|ofs| ofs + self.tx_offset())
+            .ok_or(PanicReason::OutputNotFound)?;
+
+        // halt with I/O error because tx should be serialized correctly into vm memory
+        let output = Output::from_bytes(&self.memory[offset..])?;
+
+        // amount isn't checked because we are allowed to send zero balances with a message
+        if !matches!(output, Output::Message { recipient, .. } if recipient == Address::zeroed()) {
+            return Err(PanicReason::NonZeroMessageOutputRecipient.into());
+        }
+
+        // validations passed, perform the mutations
+
+        // subtract the balance only if there is an amount set
+        if amount > 0 {
+            if self.is_external_context() {
+                self.base_asset_balance_sub(amount)?;
+            } else {
+                self.external_asset_balance_sub(amount)?;
```

Comment:
The SMO amount only ever refers to the base asset. In ether context, `self.base_asset_balance_sub` should be used

Reply:
https://github.com/FuelLabs/fuel-vm/pull/159/commits/4398216d539b52c63b2098ea4ddf38245b480924

Do you think we should rephrase here to simplify the description since we only update the base asset?

https://github.com/FuelLabs/fuel-specs/blob/bd6ec935e3d1797a192f731dadced3f121744d54/specs/vm/instruction_set.md?plain=1#L1417

Reply:
That line does say `balanceOfStart(0)` and `asset ID 0`, so it's explicit and unambiguous if you know what you're looking for. More generally, the specs aren't approachable _unless_ you've read them almost entirely, and that probably should be improved.